### PR TITLE
FM2-156:Update Patient DAO and Service

### PR DESCRIPTION
**Pull request by HerbertYiga**
https://issues.openmrs.org/browse/FM2-156

Description

Update Patient DAO and Service

### CODE REVIEW

**Comment by sherrif10** [comment]
Thanks herbert for taking this up, You can also change in tests classes like how you did here but on only person classes

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirPatientDaoImpl.java [hunk 1]
```diff
@@ -41,7 +43,7 @@ public Patient getPatientById(Integer id) {
 	}
 	
 	@Override
-	public Patient getPatientByUuid(String uuid) {
+	public Patient get(String uuid) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirPatientDaoImpl.java:46]: Technically, you should be able to remove this entire method, as its implemented in `BaseFhirDao` now.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java [hunk 2]
```diff
@@ -41,8 +43,8 @@
 	
 	@Override
 	@Transactional(readOnly = true)
-	public Patient getPatientByUuid(String uuid) {
-		return translator.toFhirResource(dao.getPatientByUuid(uuid));
+	public Patient get(String uuid) {
+		return translator.toFhirResource(dao.get(uuid));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: Technically, you should be able to remove this entire method as it's implemented in `BaseFhirService` now.

Comment (HerbertYiga) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: hi @ian  did you mean the get() for this(String uuid)?

Comment (HerbertYiga) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: @ibacher hei ian,what do you think of classess like FhirPatientDaoImpl,it extends BasePersonDao and makes use of its methods,are we dropping it,if so what about some of its methods that we are using??thanks

Comment (VaishSiddharth) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: I don't think that you need to drop `BasePersonDao` obviously we need methods like `searchForPatients` etc. I guess @ibacher meant to say that you should remove `getPatientByUuid` method from Dao as well as from the Service class as the `get` method is already implemented in the `BaseFhirService`

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: Yes, what @VaishSiddharth said. The simple get, create, update, delete methods should all be implemented by BaseFhirDao. The addition stuff shared among person objects (Person, Provider, Patient, User) should still live in BasePersonDao.

Comment (HerbertYiga) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: sicne FhirPatientDaoImpl extends BasePersonDao arleady,i am wondering on the work round to have the two extended ie plus BaseFhirDao. cc  @ibacher @VaishSiddharth

Comment (VaishSiddharth) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: @HerbertYiga So if you see closely in `BasePersonDao` it already extends `BaseFhirDao` so no need to extend that class again. And I don't think we can actually extend two classes.

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirPatientServiceImpl.java:47]: No, you can only extend one class at a time in Java... the alternative is multiple inheritance, which [can get messy](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem). That said, the class hierarchy can be as deep as you want so `FhirPatientDaoImpl` -> `BasePersonDao` -> `BaseFhirDao` -> `BaseDao` is fine. That way, `FhirPatientDaoImpl` has all the functionality of `BaseDao` and the functionality of `BaseFhirDao` and the functionality of `BasePersonDao`.

**Comment by HerbertYiga** [comment]
@ibacher  @VaishSiddharth   got this passing localy but it looks like the CI breaks with this error (java.lang.NoClassDefFoundError: Could not initialize class org.openmrs.Obs) any thoughts?

**Comment by VaishSiddharth** [comment]
@HerbertYiga It works on my machine as well. I would suggest you to rebase this branch with master.

**Comment by HerbertYiga** [comment]
Ok thanks i really appreciate your efforts @VaishSiddharth

**Comment by HerbertYiga** [comment]
@VaishSiddharth @ibacher   rebased this as suggested,though its still breaking

**Comment by HerbertYiga** [comment]
CI fails with the same logs as previous @VaishSiddharth cc @ibacher r

**Comment by ibacher** [comment]
@HerbertYiga Thanks for this! It has been committed [here](https://github.com/openmrs/openmrs-module-fhir2/commit/f619d930c5a90f8c36b7c015b1ad76b30888cd2d).

**Pull request closed without merging**